pmldr/pmldr/fs.asm

```diff
@@ -0,0 +1,333 @@
+	mov ds, ax
+
+	call LoadDiskSectors
+	pop ebp
+	ret
+
+;Tries to locate the kernel binary (ANDY.SYS) in the root directory
+FindKernelBinary:
+	push ebp
+	mov ebp, esp
+
+	mov ax, 0xF0	;start of root dir
+	mov ds, ax
+	xor ax, ax
+	mov es, ax
+	mov ax, 0x0
+	mov fs, ax										;fs points to the origin, which our local offsets are measured to.
+
+	;Each directory entry is 32 (0x20) bytes long.  Use bx to track which one we are on
+	mov bx, 0
+
+	_next_entry:
+	cmp bx, 0xFF
+	jz _kern_not_found		;break after we've looped 255 times. FIXME - should be a bit more scientific about this!
+	mov ax, 0x20
+	mul bx
+	inc bx
+
+	mov si, ax				;the file name is 11 chars (bytes) at the start of the entry, i.e. (entry_count * 0x20)
+	mov di, KernelName		;compare this to the stored expected name
+	mov cx, 11				;compare this many chars
+	call strncmp			;should preserve bx. AX=0 on success
+	test ax, ax
+	jnz _next_entry			;we didn't find one
+
+	;if we got here, we have a match! We can extract the length and starting location.
+	;Unfortunately we broke our pointer, so must reconstruct it.
+	dec bx
+	mov ax, 0x20
+	mul bx
+	mov si, ax		;ok, so si should now point to the start of the directory entry we found.
+	mov ax, word [ds:si+0x1A]	;start of the file in clusters (low word, for FAT32) is at offset 1A
+	mov di, KernelClusterLocationLow
+	stosw			;rely on es still being set to 0
+	mov ax, word [ds:si+0x14]	;high word of file start location for FAT32. Only valid if we are on FAT32
+	mov di, KernelClusterLocationHigh
+	stosw
+	mov ax, word [ds:si+0x1C]	;low word of the file length in bytes
+	mov dx, word [ds:si+0x1e]	;high word of the file length in bytes
+	mov word [fs:KernelSizeInBytesPtr], ax	;store the low word for when we need it later
+	mov word [fs:KernelSizeInBytesPtr+2], dx ;same for the high word
+	mov cx, BootSectorMemSectr
+	mov ds, cx
+	mov si, BytesPerSector
+	mov bx, word [ds:si]
+	div bx						;divide total number of bytes by bytes per sector, to get the length in sectors (stored in ax)
+	inc ax 						;assume that there is some remainder so grab another sector to be safe
+	mov di, KernelSectorLength
+	stosw
+
+	pop ebp
+	ret
+
+	_kern_not_found:
+	mov si, NotFoundErrString
+	sub si, 0x7E00
+	call PrintString
+	hlt
+	jmp $
+
+;Actually loads the kernel binary, based on the information we derived in FindKernelBinary.
+LoadInKernel:
+	push ebp
+	mov ebp, esp
+
+	mov ax, 0x50
+	mov ds, ax
+	mov gs, ax
+	mov ax, 0x0
+	mov fs, ax										;fs points to the origin, which our local offsets are measured to.
+	call F32GetRootDirLocation	
+	;root dir is at cluster 2, this location is now in ax. Subtract two clusters worth from it.
+	mov bx, word [gs:CustomDiskBlocksPerCluster]
+	sub ax, bx
+	sub ax, bx 
+	mov word [fs:DiskDataBlocksOffset], ax
+
+	mov dx, 0										;Use dx to track the number of sectors loaded
+	mov ax, 0x1000									;start of write buffer
+	mov ds, ax
+	mov ax, word [fs:KernelClusterLocationLow]			;FIXME - upper word ignored in FAT32!
```

Comment:
FIXME: we should be able to use EAX in LoadDiskSectors, it should be updated to use that